### PR TITLE
Clean up drill-through menus and add View Details action to unaggregated rows

### PR DESCRIPTION
**Pull request by akiselev**
See the [project doc](https://www.notion.so/metabase/Clean-up-and-add-view-details-action-to-drill-through-menus-on-unaggregated-rows-63cd958ec5884127a991b7af1d9435b3)

**To Reproduce**
1. Browse Data > Sample Database > Review
2. All cell filter popovers should include a `View Details` option
 image
4. Categories and short text fields should show the `Is {value}`/`Is not {value}` shortcuts instead of `=`/`!=` filters
 image
6. Long text fields should show `Contains...`/`Does not contain...` filters
 image
7. The colors in filter popovers and column header filter popovers should be consistent (b&w, with brand & white on hover)
 image

### CODE REVIEW

**Comment by ranquild** [comment]
I wasn't able to fully review this (yet) but the fact that mbql clauses were made optional in some places makes me think that this would break existing features. For anyone else reviewing this, keep this in mind.

**Comment by akiselev** [comment]
> I wasn't able to fully review this (yet) but the fact that mbql clauses were made optional in some places makes me think that this would break existing features. For anyone else reviewing this, keep this in mind.

What do you mean? Which clauses?

I tried to type everything with minimal changes outside the drills folder but I might have missed something when cleaning up for PR

**Comment by kulyk** [comment]
@akiselev for future, could you please separate refactoring changes and feature changes?

Now when files are converted to TypeScript, it's very hard too see what exactly changed about the drill logic

**Comment by kulyk** [comment]
Some visual feedback

1. That one looks a bit crowded (notice how "x-ray" and "compare" buttons stick to the subheading, "View these records" also feels a bit to close the filter section). To reproduce: Reviews table > Click the "Reviewer" column heading > Distribution > click "Others" bar

CleanShot 2022-09-16 at 18 04 06@2x

2. In Figma, the FK menu doesn't have a horizontal divider between "View this Record's Items" and "View details". Also, subjectively the popover feels too large, but I'd trust @mazameli's feedback here for sure

CleanShot 2022-09-16 at 18 06 36@2x

3. It'd be nice if the popover could stick to the cursor. Try clicking Review's Body column cell from the right side. The popover would appear on the left and I'd need to go through the whole column width to get to it:

CleanShot 2022-09-16 at 18 09 31

File: frontend/src/metabase/modes/components/drill/ColumnFilterDrill.tsx [hunk 1]
```diff
@@ -3,26 +3,36 @@ import React from "react";
...
+    ) ||
     clicked.column.field_ref == null ||
     clicked.value !== undefined
   ) {
     return [];
   }
 
   const { dimension } = clicked;
-  const initialFilter = dimension.defaultFilterForDimension();
+  const initialFilter = dimension?.defaultFilterForDimension();
```
Comment (kulyk) [inline frontend/src/metabase/modes/components/drill/ColumnFilterDrill.tsx:35]: Should we offer this drill at all if a dimension is missing?

File: frontend/src/metabase/modes/components/drill/ColumnFilterDrill.tsx [hunk 2]
```diff
@@ -35,7 +45,7 @@ export default function ColumnFilterDrill({ question, clicked }) {
       popover: ({ onChangeCardAndRun, onResize, onClose }) => (
         <FilterPopover
           query={query}
-          filter={initialFilter}
+          filter={initialFilter as any}
```
Comment (kulyk) [inline frontend/src/metabase/modes/components/drill/ColumnFilterDrill.tsx:48]: TypeScript warns that `initialFilter` can be `undefined`. Should we offer this drill if we can't figure out the filter?

File: frontend/src/metabase/modes/components/drill/DashboardClickDrill.tsx [hunk 3]
```diff
@@ -33,7 +41,7 @@ export default ({ question, clicked }) => {
     return [];
   }
   const { extraData } = clicked || {};
-  const data = getDataFromClicked(clicked);
+  const data = getDataFromClicked(clicked as any);
```
Comment (kulyk) [inline frontend/src/metabase/modes/components/drill/DashboardClickDrill.tsx:44]: Should we maybe extend `ClickActionProps` type's `extraData` type with what can actually be there?
Now it's `Record<string, any>`

File: frontend/src/metabase/modes/components/drill/DashboardClickDrill.tsx [hunk 4]
```diff
@@ -57,20 +65,20 @@ export default ({ question, clicked }) => {
       behavior = {
         ignoreSiteUrl: true,
         url: () =>
-          renderLinkURLForClick(clickBehavior.linkTemplate || "", data),
+          renderLinkURLForClick(clickBehavior.linkTemplate || "", data as any),
```
Comment (kulyk) [inline frontend/src/metabase/modes/components/drill/DashboardClickDrill.tsx:68]: Could we align the types here? They seem really close, guess there's just some notation weirdness

File: frontend/src/metabase/modes/components/drill/DashboardClickDrill.tsx [hunk 5]
```diff
@@ -57,20 +65,20 @@ export default ({ question, clicked }) => {
...
+      if (extraData?.dashboard.id === targetId) {
         const parameterIdValuePairs = getParameterIdValuePairs(
           parameterMapping,
           { data, extraData, clickBehavior },
         );
 
         behavior = {
           action: () => {
-            return dispatch =>
+            return (dispatch: any) =>
```
Comment (kulyk) [inline frontend/src/metabase/modes/components/drill/DashboardClickDrill.tsx:79]: We can use `import type { Dispatch } from "metabase-types/store"` here

File: frontend/src/metabase/modes/components/drill/DashboardClickDrill.tsx [hunk 6]
```diff
@@ -57,20 +65,20 @@ export default ({ question, clicked }) => {
...
               parameterIdValuePairs.forEach(([id, value]) => {
-                setParameterValue(id, value)(dispatch);
+                (setParameterValue(id, value) as any)(dispatch);
```
Comment (kulyk) [inline frontend/src/metabase/modes/components/drill/DashboardClickDrill.tsx:81]: Can we avoid `as any` here? Or maybe add a comment explaining why is it here

File: frontend/src/metabase/modes/components/drill/DashboardClickDrill.tsx [hunk 7]
```diff
@@ -108,7 +116,10 @@ export default ({ question, clicked }) => {
         clickBehavior,
       });
 
-      const path = Urls.dataAppPage({ id: dataAppId }, { id: targetId });
+      const path = Urls.dataAppPage(
+        { id: dataAppId } as any,
+        { id: targetId } as any,
+      );
```
Comment (kulyk) [inline frontend/src/metabase/modes/components/drill/DashboardClickDrill.tsx:122]: Please don't disable type checking

File: frontend/src/metabase/modes/components/drill/SummarizeColumnDrill.ts [hunk 8]
```diff
@@ -60,7 +69,7 @@ export default ({ question, clicked = {} }) => {
           .aggregate([aggregator.short, fieldRefForColumn(column)])
           .question()
           .setDefaultDisplay(),
-      action: () => dispatch => {
+      action: () => (dispatch: any) => {
```
Comment (kulyk) [inline frontend/src/metabase/modes/components/drill/SummarizeColumnDrill.ts:72]: We can use `import type { Dispatch } from "metabase-types/store"` here

**Review by kulyk — COMMENTED** [review]
Nice work and thanks for going the extra mile to convert drills to TypeScript. I'm a bit worried about the number of `any` this PR introduces, i.e. it's not clear if we're actually going to benefit from TypeScript here.

I'd really appreciate it if you could break this PR down at least into two (JS > TS, feature changes). I noticed a few drills were heavily changed and it's really hard to follow the diff now

File: frontend/src/metabase/modes/components/drill/DashboardClickDrill.tsx (hunk 7 above)
Comment (kulyk) [reply frontend/src/metabase/modes/components/drill/DashboardClickDrill.tsx:122]: Oh, I now see it didn't make sense from types perspective before that, my bad. In any case, we should really avoid adding `as any` all around here

Comment (akiselev) [reply frontend/src/metabase/modes/components/drill/DashboardClickDrill.tsx:122]: My original solution was to change `dataAppPage` argument annotations to `Pick<..., "id">` which keeps information in the annotation without restricting the type so much, but I changed to the `any` because of a conflict with some data app changes.

What do you think about using `Pick`?

File: frontend/src/metabase/modes/components/drill/ColumnFilterDrill.tsx (hunk 1 above)
Comment (akiselev) [reply frontend/src/metabase/modes/components/drill/ColumnFilterDrill.tsx:35]: There's a web of assumptions in the drill through components that started clashing when I added annotations. 

In this case it shouldn't happen but when the filter popover gets an `undefined` initial filter it shows the field picker. I could use an assertion to make sure that `dimension` is not undefined and throw, but that'll go through the React error boundary and show an error, which I think is worse from a UX perspective.

**Comment by akiselev** [comment]
> I'd really appreciate it if you could break this PR down at least into two (JS > TS, feature changes). I noticed a few drills were heavily changed and it's really hard to follow the diff now

I'll go ahead and separate them out

File: frontend/src/metabase/modes/components/drill/DashboardClickDrill.tsx (hunk 7 above)
Comment (kulyk) [reply frontend/src/metabase/modes/components/drill/DashboardClickDrill.tsx:122]: `Pick` sounds great to me. The reason we're passing full objects is that we often need a `name` property to build a slug as well, but there are many cases when we need a URL, and we don't have the full object that'd suit TypeScript. `Pick` sounds as a way to go here 👍

File: frontend/src/metabase/modes/components/drill/ColumnFilterDrill.tsx (hunk 1 above)
Comment (kulyk) [reply frontend/src/metabase/modes/components/drill/ColumnFilterDrill.tsx:35]: Should we maybe just return `[]` then not to offer the filter at all?

**Review by mazameli — CHANGES_REQUESTED** [review]
This doesn't appear to work on cells in Models:

image

image

And yeah, as Anton mentioned, there's some work to be done to get icon sizing, and overall spacing and padding to match the specs, e.g. this example (spec on right):

image
